File: SVN_REPO/assign3/kotlin/src/main/kotlin/meetingdsl/scheduler.kt
package meetingdsl

import java.time.LocalDate

data class Time(val hours: Int, val minutes: Int) {
    override fun toString(): String {
        return "$hours hours $minutes minutes"
    }
}

data class MeetingDetails(
  var name: String = "",
  var start: IntRange = 0..0,
  var end: IntRange = 0..0,
  var date: LocalDate? = null, 
  var participants: List<String> = emptyList()) 
{
  fun duration(): Time {
    val totalMinutesStart = start.start * 60 + start.endInclusive
    val totalMinutesEnd = end.start * 60 + end.endInclusive
    val diffMinutes = totalMinutesEnd - totalMinutesStart

    return Time(diffMinutes / 60, diffMinutes % 60)
  }

  fun generateMeetingString(): String {
    val numberOfParticipants = participants.size
    return buildString {
        append("Meeting: $name")
        if (start != 0..0) {
            append(", Start Time: ${start.start}:${start.endInclusive}")
        }
        if (end != 0..0) {
            append(", End Time: ${end.start}:${end.endInclusive}")
        }
        if (start != 0..0 && end != 0..0) {
            append(", Duration: ${duration()}")
        }
        if (date != null) {
            append(", Date: ${dateAsString()}")
        }
        if (numberOfParticipants > 0) {
            append(", Participants: ${participants.joinToString(", ")} ($numberOfParticipants participants)")
        }
    }
  }

  fun dateAsString(): String {
    return "${date?.monthValue}/${date?.dayOfMonth}/${date?.year}"
  }
}

object schedule {
  infix fun meeting(closure: MeetingDetails.() -> Unit): String {
    val details = MeetingDetails().apply(closure)
    return details.generateMeetingString()
    }
}

File: SVN_REPO/assign3/ruby/src/scheduler.rb
module MeetingDSL
  class MeetingDetails
    Time = Struct.new(:hours, :minutes) do
      def to_s
        "#{hours} hours #{minutes} minutes"
      end
    end

    attr_accessor :name, :start, :end, :date, :participants

    def initialize
      @name = ''
      @start = nil
      @end = nil
      @date = nil
      @participants = []
    end

    def name(value)
      @name = value
      self
    end

    def start(value)
      @start = value
      self
    end

    def end(value)
      @end = value
      self
    end

    def date(month, day, year)
      @date = "#{month}/#{day}/#{year}"
      self
    end

    def participants_list(value)
      @participants = value
      self
    end

    def duration
      return Time.new(0, 0) unless @start && @end

      total_minutes_start = @start.begin * 60 + @start.end
      total_minutes_end = @end.begin * 60 + @end.end
      diff_minutes = total_minutes_end - total_minutes_start
      Time.new(diff_minutes / 60, diff_minutes % 60)
    end

    def generate_meeting_string
      result = "Meeting: #{@name}"
      result += ", Start Time: #{@start.begin}:#{@start.end}" if @start
      result += ", End Time: #{@end.begin}:#{@end.end}" if @end
      result += ", Duration: #{duration}" if @start && @end
      result += ", Date: #{@date}" if @date
      result += ", Participants: #{participants.join(', ')} (#{participants.size} participants)" unless @participants.empty?
      result
    end
  end
end

module Schedule
  def self.meeting(&block)
    details = MeetingDSL::MeetingDetails.new
    details.instance_eval(&block)
    details.generate_meeting_string
  end
end

